src/tracing/metric_macros.h

```diff
@@ -0,0 +1,177 @@
+/*
+ *
+ *    Copyright (c) 2024 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#pragma once
+
+#include <lib/core/CHIPError.h>
+#include <matter/tracing/build_config.h>
+
+#if MATTER_TRACING_ENABLED
```

Comment:
can we remove this guard? since these are macros, they should not affect compilation.

Reply:
@anush-apple ?

Reply:
I need to define a version of the MATTER_LOG_METRIC_XYZ when tracing is disabled.

Reply:
When tracing is disabled, include paths are different. This is how the current tracing works:

- dependency is in `matter_trace_config` which is defined in `tracing_args.gni`
- this config has a `-I` that specifies the include paths. There are multiple `matter/tracing/macros_impl.h` that provide different implementations.

When tracing is disabled, the `none` header is included that makes everything a noop.

Reply:
Expectation would be that https://github.com/project-chip/connectedhomeip/blob/master/src/tracing/none/include/matter/tracing/macros_impl.h defines the noop macros. Your backend(s) should define the real implementation.

We do not use ifdefs.

Reply:
I guess `we mostly do not use ifdefs`: https://github.com/project-chip/connectedhomeip/blob/master/src/tracing/macros.h#L22 has a top level ifdef.


however that one imports most of the bits it needs from other places when needed. I would propose:

- place the top level macros inside `macros.h`
- place the helper macros (args bits) inside a separate header that you include.

There seems no reason to split out `tracing` from `metric tracing`. If I trace, I seem to want all capability. We did not split out begin/end/instant/scope/counter/metric before. Lets not split now.

Reply:
I followed the stuff that was done for all the Data Logging macros since this is not tracing. This is logging metric data. Those are not defined in macros_impl.h. They are defined in macros.h. I just defined a metric_macros.h and that just needs to be included in any implementation file that need to use the metric macros.

Reply:
This is resolved